File: examples/csharp/Patterns/Mediator.cs
namespace DesignPatterns.Mediator
{
    public interface IChatMediator
    {
        void SendMessage(string message, User sender);
        void AddUser(User user);
    }

    public class ChatRoom : IChatMediator
    {
        private readonly List<User> _users = new();

        public void AddUser(User user)
        {
            _users.Add(user);
            Console.WriteLine($"  {user.Name} joined the chat");
        }

        public void SendMessage(string message, User sender)
        {
            foreach (var user in _users)
                if (user != sender)
                    user.Receive(message, sender.Name);
        }
    }

    public class User
    {
        public string Name { get; }
        private readonly IChatMediator _chat;

        public User(string name, IChatMediator chat)
        {
            Name = name;
            _chat = chat;
            _chat.AddUser(this);
        }

        public void Send(string message)
        {
            Console.WriteLine($"  {Name} sends: {message}");
            _chat.SendMessage(message, this);
        }

        public void Receive(string message, string from)
        {
            Console.WriteLine($"  {Name} received from {from}: {message}");
        }
    }

    public static class Demo
    {
        public static void Run()
        {
            var chat = new ChatRoom();
            var alice = new User("Alice", chat);
            var bob = new User("Bob", chat);
            var charlie = new User("Charlie", chat);

            alice.Send("Hi everyone!");
            bob.Send("Hey Alice!");
        }
    }
}

File: examples/cpp/mediator.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class User;

class ChatRoom {
    std::vector<User*> users_;
public:
    void addUser(User* user) { users_.push_back(user); }
    void sendMessage(const std::string& message, User* sender);
};

class User {
    std::string name_;
    ChatRoom& room_;
public:
    User(const std::string& name, ChatRoom& room) : name_(name), room_(room) {
        room_.addUser(this);
    }
    const std::string& getName() const { return name_; }
    void send(const std::string& message) {
        std::cout << name_ << " sends: " << message << std::endl;
        room_.sendMessage(message, this);
    }
    void receive(const std::string& message, const std::string& from) {
        std::cout << "  " << name_ << " received from " << from << ": " << message << std::endl;
    }
};

void ChatRoom::sendMessage(const std::string& message, User* sender) {
    for (auto* user : users_) {
        if (user != sender) {
            user->receive(message, sender->getName());
        }
    }
}

int main() {
    std::cout << "=== Mediator: Chat Room ===" << std::endl;

    ChatRoom room;
    User alice("Alice", room);
    User bob("Bob", room);
    User charlie("Charlie", room);

    std::cout << std::endl;
    alice.send("Hi everyone!");
    std::cout << std::endl;
    bob.send("Hey Alice!");

    return 0;
}
